feat(routes): Implement `/info` and `/version`

vkhitrin opened the pull request:
Initial implementation for the following routes:
- `/info` - returns container-related information.
- `/version` - returns information about the server.

Signed-off-by: Vadim Khitrin <me@vkhitrin.com>

benoitf commented on Sources/socktainer/Routes/VersionRoute.swift:
```diff
@@ -2,10 +2,32 @@ import Vapor
+                Platform: ServerPlatform(Name: "socktainer"),
+                Components: [],
+                Version: ProcessInfo.processInfo.environment["SOCKTAINER_VERSION"] ?? "unknown",
+                ApiVersion: "v1.51",
```
we could enhance as part of https://github.com/socktainer/socktainer/pull/29 the build information to have this part of the makefile or then later externalize it to a separate file

benoitf replied:
notice also that some tools expect to see a docker version and not a socktainer version ( to know if something is supported or not) so at the end we might need to provide some fake versions in standard fields

vkhitrin replied:
`Ack,` I will most likely follow #29 as well.

What API versions would we want to settle on?

benoitf commented on Sources/socktainer/Routes/VersionRoute.swift:
```diff
@@ -2,10 +2,32 @@ import Vapor
+                Version: ProcessInfo.processInfo.environment["SOCKTAINER_VERSION"] ?? "unknown",
+                ApiVersion: "v1.51",
+                MinAPIVersion: "v1.51",
+                GitCommit: ProcessInfo.processInfo.environment["GIT_COMMIT"] ?? "unknown",
```
as part of https://github.com/socktainer/socktainer/pull/29 I've brought a BuildInfoHelper with getBuildCommit if that works for you

vkhitrin replied:
Thank you!

benoitf commented:
FYI @vkhitrin , once /info and /version are provided it's possible to run a basic testcontainers example

I've added a sample there: https://github.com/socktainer/sample-testcontainers-java

benoitf commented on Sources/socktainer/Routes/InfoRoute.swift:
```diff
@@ -6,6 +7,47 @@ struct InfoRoute: RouteCollection {
+                Name: hostName(),
+                ExperimentalBuild: true,
+                ServerVersion: try await ClientHealthCheck.ping().apiServerVersion,
+                // TODO: Derive this dynamically somehow
```
runtime with apple container is per container. it might be ok to return an empty array for Runtimes

vkhitrin replied:
Please ignore my comment above, it is not applicable.
I was thinking about the Kernel...

vkhitrin commented on Sources/socktainer/Routes/VersionRoute.swift:
```diff
@@ -2,10 +2,32 @@ import Vapor
+                Arch: "arm64",
+                KernelVersion: getKernel(),
+                Experimental: isDebug(),
+                BuildTime: ProcessInfo.processInfo.environment["SOCKTAINER_BUILD_TIME"] ?? "unknown",
```
Thank you!

vkhitrin commented:
Thanks, introduced the functions from #29, and extended using similar logic to publish
_hard-codded_ definitions of supported Docker Engine API versions.

Removed references to runtime.

Testing:

```
docker info
Client:
 Version:    28.4.0
 Context:    default
 Debug Mode: false

Server:
 Containers: 2
  Running: 1
  Paused: 0
  Stopped: 1
 Images: 6
 Server Version: container-apiserver version 0.4.1 (build: release, commit: 4ac18b5)
 Plugins:
  Volume:
  Network:
  Log:
 Swarm:
  NodeID:
  Is Manager: false
  Node Address:
 Init Binary:
 containerd version:
 runc version:
 init version:
 Kernel Version: vmlinux-6.12.28-153
 OSType: linux
 Architecture: arm64
 CPUs: 12
 Total Memory: 48GiB
 Name: Vadims-MacBook-Pro.local
 Docker Root Dir: /Users/vkhitrin/Library/Application Support/com.apple.container
 Debug Mode: true
  File Descriptors: 0
  Goroutines: 0
  System Time: 2025-09-21T22:47:15.454Z
  EventsListeners: 0
 Experimental: true
 Live Restore Enabled: false
 Product License: socktainer [Apache License 2.0], Apple container [Apache License 2.0]

WARNING: Apple container system info may differ
NOTE: socktainer is still under active development and considered experimental!
```

```
docker version
Client:
 Version:           28.4.0
 API version:       1.51
 Go version:        go1.24.7
 Git commit:        d8eb465
 Built:             Wed Sep  3 20:56:26 2025
 OS/Arch:           darwin/arm64
 Context:           default

Server: socktainer
 Engine:
  Version:          0.0.0-dev
  API version:      1.51 (minimum version 1.51)
  Go version:
  Git commit:       233256e
  Built:            Sun Sep 21 22:46:09 2025
  OS/Arch:          macOS/arm64
  Experimental:     true
```

benoitf reviewed (commented):
checking against the testcontainers sample we have



```
ava.lang.IllegalStateException: Check failed: Docker server version should be at least 1.6.0
        at org.testcontainers.DockerClientFactory.check(DockerClientFactory.java:280)
        at org.testcontainers.DockerClientFactory.checkDockerVersion(DockerClientFactory.java:272)
        at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:257)
        at org.testcontainers.DockerClientFactory$1.getDockerClient(DockerClientFactory.java:108)
        at com.github.dockerjava.api.DockerClientDelegate.authConfig(DockerClientDelegate.java:109)
        at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:321)
        at com.example.testcontainers.PostgreSQLServiceTest.testPostgreSQLSample(PostgreSQLServiceTest.java:31)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
```

so we need to provide a fake docker version there, not the version of socktainer for now

It's possible to provide socktainerVersion through its own new property

for example podman, adds a

```
 "BuildahVersion": "1.40.1",
```

entry

you can check on your own using https://github.com/socktainer/sample-testcontainers-java

and running `TESTCONTAINERS_RYUK_DISABLED=true mvn install`

vkhitrin commented on Package.swift:
```diff
@@ -13,6 +13,7 @@ let package = Package(
     ],
     dependencies: [
         .package(url: "https://github.com/apple/container.git", from: "0.4.1"),
+        .package(url: "https://github.com/apple/containerization.git", from: "0.5.0"),
```
Thanks, will bump this version!

vkhitrin commented on Sources/BuildInfo/include/BuildInfo.h:
```diff
@@ -10,8 +10,20 @@
 #define BUILD_TIME "unspecified"
 #endif
 
+#ifndef DOCKER_ENGINE_API_MIN_VERSION
+#define DOCKER_ENGINE_API_MIN_VERSION "1.51"
+#endif
+
+#ifndef DOCKER_ENGINE_API_MAX_VERSION
+#define DOCKER_ENGINE_API_MAX_VERSION "1.51"
+#endif
+
```
Would we want this field be configurable dynamically during build time?

benoitf replied:
yes I think it's part of the build/release step

also maybe I think minimum version should be lowered for some other clients, they should still be able to query previous versions

vkhitrin replied:
Should we set for `v1.32`, as I saw this is the version used in the testcontainer's SDK in the example project.?

vkhitrin commented:
This is interesting, when the version is `unknown` (compiling directly using `swift build`), the testcontainer passes:

```
docker version
Client:
 Version:           28.4.0
 API version:       1.51
 Go version:        go1.24.7
 Git commit:        d8eb465
 Built:             Wed Sep  3 20:56:26 2025
 OS/Arch:           darwin/arm64
 Context:           default

Server: socktainer
 Engine:
  Version:          unspecified
  API version:      1.51 (minimum version 1.51)
  Go version:
  Git commit:       unspecified
  Built:            unspecified
  OS/Arch:          macOS/arm64
  Experimental:     true
```

Running the example project:

```
.....
# From log
13:24:32.848 [main] DEBUG o.testcontainers.DockerClientFactory - Docker version: {Components=[], KernelVersion=25.0.0, ApiVersion=1.51, Platform={Name=socktai
ner}, GitCommit=unspecified, Os=macOS, Version=unspecified, MinAPIVersion=1.51, Experimental=true, Arch=arm64, BuildTime=unspecified}
13:24:32.848 [main] INFO  o.testcontainers.DockerClientFactory - Connected to docker:
  Server Version: container-apiserver version 0.4.1 (build: release, commit: 4ac18b5)
  API Version: 1.51
  Operating System: null
  Total Memory: 49152 MB
....
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
.....
```

When setting the dev build:

```
docker version
Client:
 Version:           28.4.0
 API version:       1.51
 Go version:        go1.24.7
 Git commit:        d8eb465
 Built:             Wed Sep  3 20:56:26 2025
 OS/Arch:           darwin/arm64
 Context:           default

Server: socktainer
 Engine:
  Version:          0.0.0-dev
  API version:      1.51 (minimum version 1.51)
  Go version:
  Git commit:       39cb240
  Built:            Mon Sep 22 10:30:23 2025
  OS/Arch:          macOS/arm64
  Experimental:     true
```

```
13:31:14.825 [main] ERROR o.testcontainers.DockerClientFactory - ❌ Docker server version should be at least 1.6.0
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.218 s <<< FAILURE! -- in com.example.testcontainers.PostgreSQLServiceTest
[ERROR] com.example.testcontainers.PostgreSQLServiceTest.testPostgreSQLSample -- Time elapsed: 0.211 s <<< ERROR!
java.lang.IllegalStateException: Check failed: Docker server version should be at least 1.6.0
        at org.testcontainers.DockerClientFactory.check(DockerClientFactory.java:280)
        at org.testcontainers.DockerClientFactory.checkDockerVersion(DockerClientFactory.java:272)
        at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:257)
        at org.testcontainers.DockerClientFactory$1.getDockerClient(DockerClientFactory.java:108)
        at com.github.dockerjava.api.DockerClientDelegate.authConfig(DockerClientDelegate.java:109)
        at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:321)
        at com.example.testcontainers.PostgreSQLServiceTest.testPostgreSQLSample(PostgreSQLServiceTest.java:31)
        at java.base/java.lang.reflect.Method.invoke(Method.java:569)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   PostgreSQLServiceTest.testPostgreSQLSample:31 » IllegalState Check failed: Docker server version should be at least 1.6.0
```

How should we proceed.

benoitf commented:
I think providing a valid semver version will make less failures from clients

maybe pickup the version of the API ?

vkhitrin commented:
>I think providing a valid semver version will make less failures from clients
> 
> maybe pickup the version of the API ?

Correct if me I am wrong, in Podman the version is set to match podman, thus it works correctly:
https://github.com/containers/podman/blob/167102951749ebe140de76bd049a4df53f437a7c/version/version.go#L28
https://github.com/containers/podman/blob/167102951749ebe140de76bd049a4df53f437a7c/version/rawversion/version.go#L7
https://github.com/containers/podman/blob/167102951749ebe140de76bd049a4df53f437a7c/libpod/define/version.go#L51

So for the current time, using `socktainer`'s version or Apple container's is counter productive.

What are our options in this case?

benoitf commented:
let's use the maximum API version and see if we have reports on this, for podman it's basically working because now they have a version greater than 1.x (now 5.x)

benoitf commented:
I think testcontainers should maybe have checked the API level version but that's another story

benoitf commented:
thanks, with other previous PRs it'll be possible to use testcontainers (at least a basic usage)
